scalameta/src/main/scala/scala/meta/internal/parsers/Parsers.scala

```diff
@@ -562,6 +562,18 @@ private[meta] class Parser(val input: Input)(implicit val dialect: Dialect) { pa
     }
   }
 
+  // corresponding quasiquote (rank 1) for expressions like `{..$smth}`
+  def constructR1Quasi[T <: Tree : AstMetadata](ctor: => Tree): List[T] = {
```

Comment:
Why's the current token positioned at `$enumerators`? From what I can see, you're explicitly rewinding the token stream before calling `constructR1Quasi`, but if you didn't do that, then you wouldn't have to special-case at least this difference with `ellipsis`.

I understand that this is done to skip `case` and `@` tokens in the case of pattern matching cases and annotations, but how about we do this differently? `ellipsis` could gain additional parameter that says which tokens it needs to skip, so that we rewrite:

```
if (token.is[Ellipsis]) {
  nextTwice()
  annots ++= constructR1Quasi[Mod.Annot](expr())
} else {
``` 

As something like:

```
if (token.is[Ellipsis]) {
  annots += ellipsis(1, skip = `@`)(expr())
} else {
```